core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,217 @@
+    Preconditions.checkArgument(ref != null, "No such tag with name %s", name);
+    Preconditions.checkArgument(ref.isTag(), "Ref with name %s is a branch not a tag", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation renameBranch(String name, String newName) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    Preconditions.checkNotNull(newName, "Branch name cannot be null");
+    Preconditions.checkArgument(!name.equals(SnapshotRef.MAIN_BRANCH), "Cannot rename main branch");
+    SnapshotRef ref = updatedRefs.get(name);
+    Preconditions.checkArgument(ref != null, "No such ref with name %s", name);
+    Preconditions.checkArgument(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    SnapshotRef existing = updatedRefs.put(newName, ref);
+    ValidationException.check(existing == null, "Ref %s already exists", newName);
+    updatedRefs.remove(name, ref);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation replaceBranch(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    Preconditions.checkArgument(ref != null, "Branch %s does not exist", name);
+    Preconditions.checkArgument(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    SnapshotRef updatedRef = SnapshotRef.builderFrom(ref, snapshotId).build();
+    updatedRefs.put(name, updatedRef);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation replaceBranch(String name, String source) {
+    return replaceBranch(name, source, false);
+  }
+
+  public UpdateSnapshotReferencesOperation fastForward(String name, String source) {
+    return replaceBranch(name, source, true);
+  }
+
+  private UpdateSnapshotReferencesOperation replaceBranch(String name, String source, boolean fastForward) {
+    Preconditions.checkNotNull(name, "Target branch cannot be null");
+    Preconditions.checkNotNull(source, "Source ref cannot be null");
+    SnapshotRef sourceRef = updatedRefs.get(source);
+    SnapshotRef refToUpdate = updatedRefs.get(name);
+    Preconditions.checkArgument(sourceRef != null, "Ref %s does not exist", source);
+    Preconditions.checkArgument(refToUpdate != null, "Branch %s does not exist", name);
+    Preconditions.checkArgument(refToUpdate.isBranch(), "Ref with name %s is a tag not a branch", name);
+
+    // Nothing to replace
+    if (sourceRef.snapshotId() == refToUpdate.snapshotId()) {
+      return this;
+    }
+
+    SnapshotRef updatedRef = SnapshotRef.builderFrom(refToUpdate, sourceRef.snapshotId()).build();
+
+    if (fastForward) {
+      Set<Long> sourceAncestors = Sets.newHashSet(SnapshotUtil.ancestorIds(
+          base.snapshot(sourceRef.snapshotId()), base::snapshot));
+      Preconditions.checkArgument(sourceAncestors.contains(refToUpdate.snapshotId()),
+          "Cannot perform fast-forward: %s is not an ancestor of %s", name, source);
```

Comment:
I would remove "perform" because "fast-forward" is already a verb. We don't need an extra one.

Reply:
Done